dhall/src/Dhall/Eval.hs

```diff
@@ -327,6 +338,21 @@ vField t0 k = go t0 where
   singletonVRecordLit v = VRecordLit (Dhall.Map.singleton k v)
 {-# inline vField #-}
 
+vProjectByFields :: Eq a => Env a -> Val a -> Set Text -> Val a
+vProjectByFields env t ks =
+  if null ks then
+    VRecordLit mempty
+  else case t of
+    VRecordLit kvs -> let
+      kvs' = Dhall.Map.restrictKeys kvs (Dhall.Set.toSet ks)
+      in VRecordLit kvs'
+    VPrefer l r@(VRecordLit kvs) -> let
+      ksSet = Dhall.Set.toSet ks
+      kvs' = Dhall.Map.restrictKeys kvs ksSet
+      ks' = Dhall.Set.fromSet (Data.Set.difference ksSet (Dhall.Map.keysSet kvs'))
+      in vPrefer env (vProjectByFields env l ks') (VRecordLit kvs')
```

Comment:
I suspect that this could be refactored to be tail-recursive, but I don't expect that to change a lot performance-wise.